src/model/enhance-images.ts

```diff
@@ -0,0 +1,333 @@
+	if (hasULwrapper && containsLItags) {
+		// element is an essay caption
+		return element.html;
+	}
+	return '';
+};
+
+/**
+ * decideTitle
+ * Decides if a text element should be used to title the preceding image
+ * @param {TextBlockElement} element - The text element to examine
+ * @param {Boolean} isPhotoEssay - We treat photo essays differently
+ * */
+const decideTitle = ({
+	element,
+	isPhotoEssay,
+}: {
+	element?: SubheadingBlockElement;
+	isPhotoEssay: boolean;
+}): string => {
+	// Checks if this element is a 'title' based on the convention: <h2>Title text</h2>
+	if (!element) return '';
+	if (!isPhotoEssay) return '';
+	// Extract title
+	const frag = JSDOM.fragment(element.html);
+	if (!frag || !frag.firstChild) return '';
+	const isH2tag = frag.firstChild.nodeName === 'H2';
+	if (isH2tag) {
+		// element is an essay title
+		return (frag.textContent && frag.textContent.trim()) || '';
+	}
+	return '';
+};
+
+const enhanceImage = ({
+	image,
+	caption,
+	title,
+}: {
+	image: ImageBlockElement;
+	caption?: string;
+	title?: string;
+}): ImageBlockElement => {
+	const enhancedImage = { ...image };
+	if (caption) {
+		enhancedImage.data.caption = caption;
+	}
+	if (title) {
+		enhancedImage.title = title;
+	}
+	return enhancedImage;
+};
+
+const processBuffer = ({
+	imageBuffer,
+	title,
+	caption,
+	isPhotoEssay,
+}: {
+	imageBuffer: ImageBlockElement[];
+	title?: string;
+	caption?: string;
+	isPhotoEssay?: boolean;
+}): CAPIElement[] => {
+	function cleanCaption(image: ImageBlockElement) {
+		// If a standalone caption was found we want to strip any existing captions
+		// Or if this is article is a photo essay, then we always strip captions, even if that
+		// leaves the image with no caption at all, we want that for photo essays
+		if (caption || isPhotoEssay) {
+			return {
+				...image,
+				data: {
+					...image.data,
+					caption: '',
+				},
+				displayCredit: false,
+			};
+		}
+		// No special caption was found so continue with this images defaults
+		return image;
+	}
+
+	const processed: CAPIElement[] = [];
+	let prevHalfWidth: ImageBlockElement | null;
+	imageBuffer.map(cleanCaption).forEach((image, i) => {
+		const endOfBuffer = i + 1 === imageBuffer.length;
+		switch (image.role) {
+			case 'halfWidth':
+				if (!prevHalfWidth) {
+					if (endOfBuffer) {
+						processed.push(enhanceImage({ image, caption, title }));
+					} else {
+						prevHalfWidth = image;
+					}
+				} else {
+					if (endOfBuffer) {
+						processed.push({
+							_type:
+								'model.dotcomrendering.pageElements.MultiImageBlockElement',
+							elementId: prevHalfWidth.elementId,
+							images: [prevHalfWidth, image],
+							caption,
+						});
+					} else {
+						processed.push({
+							_type:
+								'model.dotcomrendering.pageElements.MultiImageBlockElement',
+							elementId: prevHalfWidth.elementId,
+							images: [prevHalfWidth, image],
+						});
+					}
```

Comment:
Not tested, but wonder if the difference between the returned objects might be better communicated with something like:

```suggestion

						const maybeCaption = endOfBuffer? {caption} : {};
						processed.push({{
							_type:
								'model.dotcomrendering.pageElements.MultiImageBlockElement',
							elementId: prevHalfWidth.elementId,
							images: [prevHalfWidth, image],
						}, ...maybeCaption});
```

Reply:
I think the destructuring is a bit hard to parse (at least I had to squint for a bit to get it!) but I defo agree this `if` statement could/should be removed. I'll see if I can find a way to do it with less squinting otherwise I'll fallback to this 👍

Reply:
Yeah, totally not wedded to the destructuring!